org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseInitHandler.java

```diff
@@ -152,7 +153,7 @@ public InitializeResult initialize(InitializeParams param) {
 
 		Integer processId = param.getProcessId();
 		if (processId != null) {
```

Comment:
Yes, I think so. IIRC I tried it initially and it worked but then tried something else ;)
I'm going to update this change as you suggest.

Reply:
Great. `setParentProcessId` is also called in `ParentProcessWatcher` in the form : `JavaLanguageServerPlugin.getLanguageServer().setParentProcessId(..)` but I suspect you're not even using this class, so aren't affected by it, given that it's called from `JavaLanguageServerPlugin.startConnection(..)`, which you're essentially rewriting in your client.

I'll have a last look and prepare to merge this.

Reply:
We can remove org.eclipse.jdt.ls.core.internal.LanguageServer.parentProcessId and use ProcessHandle.current().parent().get().pid() when initializing ParentProcessWatcher.
``` 
diff --git a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java
index dd7aa931..57b6335d 100644
--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java
@@ -39,12 +39,11 @@ public final class ParentProcessWatcher implements Runnable, Function<MessageCon
        private final LanguageServer server;
        private ScheduledFuture<?> task;
        private ScheduledExecutorService service;
+       private long pid;
 
        public ParentProcessWatcher(LanguageServer server ) {
                this.server = server;
-               if (ProcessHandle.current().parent().isPresent()) {
-                       this.server.setParentProcessId(ProcessHandle.current().parent().get().pid());
-               }
+               this.pid = ProcessHandle.current().parent().get().pid();
                service = Executors.newScheduledThreadPool(1);
                task =  service.scheduleWithFixedDelay(this, POLL_DELAY_SECS, POLL_DELAY_SECS, TimeUnit.SECONDS);
        }
@@ -74,7 +73,7 @@ public final class ParentProcessWatcher implements Runnable, Function<MessageCon
         */
        private boolean parentProcessStillRunning() {
                // Wait until parent process id is available
-               final long pid = server.getParentProcessId();
+               // final long pid = server.getParentProcessId();
                if (pid == 0 || lastActivityTime > (System.currentTimeMillis() - INACTIVITY_DELAY_SECS)) {
                        return true;
                }
```

Reply:
I hesitate a bit because it is part of the [language spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeParams), but we only use it in the parent process watcher, and `ProcessHandle` does make it pretty obsolete. If we run into a situation where a client may require it, then we can always re-introduce.

@mickaelistria , feel free to adopt @snjeza 's suggestion in this PR.

Reply:
I also think we should keep the pid as defined in the spec and react to it. There is actually nothing in the spec that mandate that the process to watch is the parent process. If you think of some launcher (eg a bash script) that would launch the language server, then the parent process is bash, and not the client.
JDT-LS may clarify it by renaming "parentProcessId" to "clientProcessId"